### DISCUSSION TITLE
How to change mail address for user?

### DISCUSSION THREAD

**Discussion by nottheend** [firefly-iii/firefly-iii#6681] [General]
Hi,

my user account was created with AUTHENTICATION_GUARD (externally handled authentication).
I would like to receive notifications. However, my user profile doesn't have an email address associated.
I tried updating the database. In table `users` I added the address to the (previously empty) field `email`.
After I login to firefly-iii again, it prompts me with the new user creation dialog (as I just finished the installation).
Is there any better way to add my mail address?

**Comment by JC5** [comment]
Not yet, I'll write it down.

**Reply by nottheend** [reply]
Thank you! Just to be clear: Is there right now a way to update the mail address for an exisiting user with keeping all data (i.e. without new user process post installation)?

**Reply by JC5** [reply]
I don't believe so, when you're authenticating externally I'm afraid. Otherwise, it's in your profile.

**Comment by nebulade** [comment]
While debugging this for the Cloudron package, which uses the auth guard, I found a missing fallback to fetch the email header https://github.com/firefly-iii/firefly-iii/pull/6685

After this the email is at least correctly set, while for some reason it is still not set as `remote_guard_alt_email` later. Still debugging this.

**Reply by JC5** [reply]
Does that function actually do anything useful? I can't find anything specific on it.

**Reply by nebulade** [reply]
If you refer to `apache_request_headers` then I am not sure myself, I am not really a php programmer, but I copied that from a few lines above and in my tests it did return the correct header value.

**Reply by nebulade** [reply]
I have further added https://github.com/firefly-iii/firefly-iii/pull/6701 to keep the email updated if it is changed in the external proxy auth user directory.

In the user admin ui, there still seems to be a discrepancy with the username being the email:
![image](https://user-images.githubusercontent.com/646768/205667878-a01c73d1-d7b0-43be-821d-6ea32ae44515.png)

**Reply by nebulade** [reply]
Once those two changes are applied, users then receive emails on that email address.

**Reply by nebulade** [reply]
@JC5 is there any chance this can make it into the codebase or is anything blocking this at the moment?

**Reply by JC5** [reply]
I lack the time to look at this. I *really* dislike introducing server specific code into Firefly III. It implies there is a problem that I'm now solving for one use case specifically and there are *n* others waiting for me. I hope to have some time for research soon. 

As it stands, I'm disinclined to merge code for deployment specific problems but I'm ready to be convinced otherwise.

**Reply by nebulade** [reply]
I fully understand. I only used that apache specific method since it was already used a few lines above that. I can see to find out what the correct non apache specific approach here is.

For https://github.com/firefly-iii/firefly-iii/pull/6701 this is independent of that though.

**Reply by JC5** [reply]
Lol that has to go. Thanks for pointing it out.

**Reply by nottheend** [reply]
Thanks @nebulade for fixing and @JC5 for merging!
I can now successfully receive the test mail 🥳

Unfortunately, I still don't receive a mail upon automated creation of a recurring transaction 🥲

**Reply by JC5** [reply]
I don't think you're supposed to get one lol but I have to check the code to be sure.

**Reply by nottheend** [reply]
Haha, alright, thanks
I thought that I received mails in a different setup in the past whenever transactions were created automatically.
This is the best personal finance manager around 👍🏻

**Reply by nottheend** [reply]
An addition: I am using firefly-iii packed by cloudron and I discovered an option in cloudron to allow **_Masquerading_**.
After enabling, I receive reminder mails upon creation of automated transactions from firefly-iii 🥳

_Masquerading allows users and apps to send emails with an arbitrary username in the FROM address._

**Reply by gramakri** [reply]
@nottheend Usually, you don't need to enable that setting. It could be that the email address the package is configured for and the email address from which firefly is sending emals is different. Can you check what https://docs.cloudron.io/apps/#mail-from-address is set to? The address in Cloudron dashboard should match what firefly-iii is sending mails from. With that, you can disable masquerading.
